libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_spline(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    bool finish;
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    float time_start = segment[7].start_time;
+    float time_mid = segment[num_segs - 1].start_time - segment[11].start_time;
+
+    update(MIN(time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1;
+    accel += _delta_unit_1 * scurve_A1;
+
+    update(MAX(time - time_start + segment[11].start_time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_2 * (scurve_P1 - segment[11].start_pos);
+    vel += _delta_unit_2 * scurve_V1;
+    accel += _delta_unit_2 * scurve_A1;
+
+    update(MIN(segment[11].start_time - (time - time_start * 2 - time_mid + segment[11].start_time), segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_3 * (segment[11].start_pos - scurve_P1);
+    vel += _delta_unit_3 * scurve_V1;
+    accel += _delta_unit_3 * (-scurve_A1);
+
+    finish = _t > (time_mid + 2.0 * time_start);
+    finish = _t > (time_mid + 2.0 * time_start);
+
+    vel *= MIN(time_scale * 1.1f, 1.0f);
+    accel *= time_scale;
+    return finish;
+}
+
+// debugging messages
+void scurves::debug()
+{
+    hal.console->printf("\n");
+    hal.console->printf("num_segs:%u, type:%4.2f, _t:%4.2f, otj:%4.2f, jerk_max:%4.2f, accel_max:%4.2f, vel_max:%4.2f\n",
+                        (unsigned)num_segs, (double)segtype, (double)_t, (double)otj, (double)jerk_max, (double)accel_max, (double)vel_max);
+    hal.console->printf("T, Jt, J, A, V, P \n");
+    for (uint8_t i = 0; i < num_segs; i++) {
+        hal.console->printf("i:%u, T:%4.2f, Jtype:%4.2f, J:%4.2f, A:%4.2f, V: %4.2f, P: %4.2f\n",
+                            (unsigned)i, (double)segment[i].start_time, (double)segment[i].jtype, (double)segment[i].jerk_ref,
+                            (double)segment[i].start_accel, (double)segment[i].start_vel, (double)segment[i].start_pos);
+    }
+    hal.console->printf("_track x:%4.2f, y:%4.2f, z:%4.2f\n", (double)_track.x, (double)_track.y, (double)_track.z);
+    hal.console->printf("_delta_unit_1 x:%4.2f, y:%4.2f, z:%4.2f\n", (double)_delta_unit_1.x, (double)_delta_unit_1.y, (double)_delta_unit_1.z);
+    hal.console->printf("_delta_unit_2 x:%4.2f, y:%4.2f, z:%4.2f\n", (double)_delta_unit_2.x, (double)_delta_unit_2.y, (double)_delta_unit_2.z);
+    hal.console->printf("_delta_unit_2 x:%4.2f, y:%4.2f, z:%4.2f\n", (double)_delta_unit_3.x, (double)_delta_unit_3.y, (double)_delta_unit_3.z);
+    hal.console->printf("\n");
+}
+
+// straight segment implementations of pos_end, time_end, time_to_end and braking
+float scurves::pos_end_straight() const
+{
+    return segment[num_segs - 1].start_pos;
```

Comment:
Makes me a bit nervous.  If num_segs is zero this could lead to fault of some kind.

Similarly elsewhere.

Reply:
this code is gone with the re-implementation of splines.  There are still 3 other situations where we use a "-1" in the index but all three look safe because they follow non-zero constants.  E.g. "segment[SEG_ACCEL_MAX-1].end_time"